game2-safe/src/Server/Server.fs

```diff
@@ -0,0 +1,129 @@
+open System
+open System.IO
+open System.Threading.Tasks
+open System.Text
+
+open Microsoft.AspNetCore.Builder
+open Microsoft.Extensions.DependencyInjection
+open FSharp.Control.Tasks.V2
+open Giraffe
+open Saturn
+open Shared
+
+open Fable.Remoting.Server
+open Fable.Remoting.Giraffe
+open Microsoft.WindowsAzure.Storage
+open Microsoft.WindowsAzure.Storage.Table
+open Microsoft.WindowsAzure.Storage.Blob
+open Newtonsoft.Json
+open Microsoft.WindowsAzure.Storage
+open Microsoft.WindowsAzure.Storage
+
+let tryGetEnv = System.Environment.GetEnvironmentVariable >> function null | "" -> None | x -> Some x
+let publicPath = tryGetEnv "public_path" |> Option.defaultValue "../Client/public" |> Path.GetFullPath
+let storageAccount = tryGetEnv "STORAGE_CONNECTIONSTRING" |> Option.defaultValue "UseDevelopmentStorage=true" |> CloudStorageAccount.Parse
+
+let blobClient = storageAccount.CreateCloudBlobClient()
+let guid = "151be5b4-3cb7-45b7-8737-65e08e8faf63"
+let blobRef = blobClient.GetContainerReference guid
+
+let blob =
+    task {
+        let! _ = blobRef.CreateIfNotExistsAsync ()
+        do! blobRef.SetPermissionsAsync
+                (BlobContainerPermissions(
+                    PublicAccess = BlobContainerPublicAccessType.Blob
+                ))
+        let file =
+            blobRef.GetBlockBlobReference "scores.json"
+        let! exists = file.ExistsAsync()
+        if not exists then
+            do! file.UploadTextAsync "[]"
+        return file
+    }
+    |> Async.AwaitTask
+    |> Async.RunSynchronously
+
+let jsonConverter = Fable.Remoting.Json.FableJsonConverter () :> JsonConverter
+
+let leaseTime = TimeSpan.FromSeconds 15.
+
+let getScores () =
+    task {
+        let b = Array.zeroCreate 65536
+        let! count = blob.DownloadToByteArrayAsync(b, 0)
+        let json = Encoding.UTF8.GetString(b, 0, count)
+        return JsonConvert.DeserializeObject<HighScores>(json, [|jsonConverter|])
+    }
+
+let addScore (score : string * int) =
+    task {
+        let! lease = blob.AcquireLeaseAsync(Nullable.op_Implicit leaseTime, null)
```

Comment:
yes it's to enable concurrent writes - I read the scores from file and append a new one. I think we're ok with the lock here as long as it's implemented correctly. Will this call block until another lease is released?

Reply:
This won't allow anyone else to acquire a lock until you release it - it's essentially a form of pessimistic concurrency. I'm not sure what'll happen if you try to get a lock whilst it's already there i.e. does it throw an exception or return null or what?

Leases aren't (to my mind) normally used for this; normally the etag approach is more common in my experience.

See [here](https://azure.microsoft.com/en-us/blog/managing-concurrency-in-microsoft-azure-storage-2/) and [here](https://cann0nf0dder.wordpress.com/2015/09/07/azure-blob-storage-and-managing-concurrency/) for a couple of examples.

Reply:
The alternative option you could do is to have this using an [Append Blob](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs). In this model, instead of you reading the blob into memory and effectively updating it, you simply append the new row to the blob. You would need to change the way you serialize / deserialize the data though - each block would probably need to be a separate json object and then you'd splice them all together in F#.